graph.c

```diff
@@ -0,0 +1,350 @@
+void graph_list(Graph *self) {
+    //allocate mem
+    GList* glist = g_hash_table_get_keys(self->vertices_table);
+
+    while(glist != NULL) {
+        printf("%d\n", *((int *)(glist->data)));
+        glist = glist->next;
+    }
+
+    //dellocate mem
+    g_list_free(glist);
+}
+
+static void graph_edge_function_check_pararameters(Graph* self, Graph_vertice* src, Graph_vertice* dest, int from_vertice, int to_vertice, int weight) {
+    if (graph_is_null_pointer(self)) {
+        fprintf(stderr, "graph pointer is equal to NULL");
+        abort();
+    }
+
+    if((src == NULL) || dest == NULL ) {
+        fprintf(stderr, "from_vertice or to_vertice does not exist in graph");
+        abort();
+    }
+
+    if (from_vertice == to_vertice) {
+        fprintf(stderr, "Self loops in graph is not allowed");
+        abort();
+    }
+
+    if(from_vertice < 0 || to_vertice < 0) {
+        fprintf(stderr, "Vertices indexes should be >= 0");
+        abort();
+    }
+
+    if (weight <= 0) {
+        fprintf(stderr, "Weight should be > 0");
+        abort();
+    }
+}
+
+static void graph_add_new_neighbor(GPtrArray *neighbors, int to_vertice, int weight) {
+    Graph_vertice_neighbor* new_neighbor;
+
+    if ((new_neighbor = (Graph_vertice_neighbor *) malloc(sizeof(*new_neighbor))) == NULL) {
+        fprintf(stderr, "memalloc for Graph_vertice_neighbor failed");
+        abort();
+    }
+
+    new_neighbor->index = to_vertice;
+    new_neighbor->weight = weight;
+
+    g_ptr_array_add(neighbors, new_neighbor);
+}
+
+static void graph_add_neighbors_table(Graph_vertice *self) {
+    GPtrArray *ptr_array = g_ptr_array_new_with_free_func(g_free);
+
+    self->neighbors = ptr_array;
+}
+
+static Graph_vertice_neighbor* graph_find_neighbor(GPtrArray *neighbors, int index) {
+    guint i;
+    Graph_vertice_neighbor* self;
+
+    for (i = 0; i < neighbors->len; i++) {
+        self = g_ptr_array_index(neighbors, i);
+        if(self->index == index) {
+            return self;
+        }
+    }
+
+    return NULL;
+}
+
+
+//add a new weighted directed edge to the graph that connects two vertices
+void graph_add_edge(Graph* self, int from_vertice, int to_vertice, int weight) {
+    Graph_vertice* src;
+    Graph_vertice* dest;
+
+    src = graph_get_vertice(self, from_vertice);
+    dest = graph_get_vertice(self, to_vertice);
+
+    graph_edge_function_check_pararameters(self, src, dest, from_vertice, to_vertice, weight);
+
+    if (dest->neighbors != NULL) {
+        if (dest->neighbors->len > 0) {
+            Graph_vertice_neighbor* n;
+
+            if((n = graph_find_neighbor(dest->neighbors, from_vertice)) != NULL) {
+                n = NULL;
+
+                fprintf(stderr, "Reverse path is found in to_vertice");
+                abort();
+            }
+
+            n = NULL;
+        }
+    }
+
+    if (src->neighbors != NULL) {
+        if (src->neighbors->len > 0) {
+            Graph_vertice_neighbor* n;
```

Comment:
you can declare an assign: `Graph_vertice_neighbor* n = nullptr;`

Reply:
nullptr is a C++ concept afaik but I got the point

Reply:
nullptr is available since C23